File: frontend/src/features/banners/types/banner.ts
// Banner Response Types
export interface BannerResponse {
  id: number;
  title?: string;
  subtitle?: string;
  imageUrl: string;
  redirectUrl?: string;
  altText?: string;
  buttonText?: string;
  displayOrder?: number;
  isActive: boolean;
  createdAt?: string;
  updatedAt?: string;
}

// Banner Request Types
export interface BannerRequest {
  title?: string;
  subtitle?: string;
  redirectUrl?: string;
  altText?: string;
  buttonText?: string;
  displayOrder?: number;
  isActive?: boolean;
}

File: backend/src/main/java/app/store/entity/Banner.java
package app.store.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import lombok.*;
import lombok.experimental.FieldDefaults;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Banner extends BaseEntityLong {
    @Column(nullable = false)
    String title;

    String subtitle;

    @Column(nullable = false, columnDefinition = "TEXT")
    String imageUrl;

    String redirectUrl;

    String altText;

    String buttonText;

    @Column(nullable = false)
    @Builder.Default
    Integer displayOrder = 0;

    @Column(nullable = false)
    @Builder.Default
    Boolean isActive = true;

    // BannerMapper map request qua builder của Lombok, mà builder luôn gọi
    // .displayOrder(...)/.isActive(...) kể cả khi request để trống -> giá trị
    // khởi tạo của @Builder.Default bị ghi đè bằng null. Chốt lại ở đây để
    // không phụ thuộc vào việc mapper có truyền null hay không.
    @PrePersist
    @PreUpdate
    void applyDefaults() {
        if (displayOrder == null)
            displayOrder = 0;
        if (isActive == null)
            isActive = true;
    }
}
